Use newtypes for Bool solver options

grayjay opened the pull request:
This commit shouldn't change any behavior, except possibly the serialization of GenericInstallPlan, InstallFlags, etc.

23Skidoo commented:
LGTM.

We may want to add a type class for such types, something like

```
class BooleanFlag a where
    asBool :: a -> Bool
```

and use `asBool` instead of `unReorderGoals`/`unAvoidReinstalls`/[...].

kosmikus commented on cabal-install/Distribution/Client/Dependency/Types.hs:
```diff
@@ -106,9 +113,30 @@ instance Text PreSolver where
+
+newtype StrongFlags = StrongFlags { unStrongFlags :: Bool }
+  deriving (Eq, Generic, Show)
+
```
If all these have an extractor, and `Eq`, `Generic`, and `Show`, shouldn't then `EnableBackjumps` get the same treatment? (Yes, it's an option not exposed to the user, but still ...)

kosmikus commented:
Looks good to me as well.

grayjay commented:
I added a `BooleanFlag` type class in the last commit.  I think it would be more useful if we moved it to the library and used it to generalize the functions that handle boolean options, but that would be a bigger change.  I created a Bool instance and used GeneralizedNewtypeDeriving to reduce boilerplate.  Does that make sense, or would it be better to write out the instances?

I also derived Eq, Show, Generic, and BooleanFlag for EnableBackjumps.

23Skidoo commented:
Great, I'll merge this PR once the build bot is green.

23Skidoo commented:
>  I created a Bool instance and used GeneralizedNewtypeDeriving to reduce boilerplate. Does that make sense, or would it be better to write out the instances?

GHC extensions (except Template Haskell) are allowed, provided that they are available on all GHC versions we need to support. Using `GeneralizedNewtypeDeriving` is certainly fine.

>  I think it would be more useful if we moved it to the library and used it to generalize the functions that handle boolean options

The right place for `BooleanFlag` is probably in `Distribution.Simple.Setup`, near the `Flag` type:
https://github.com/haskell/cabal/blob/master/Cabal/Distribution/Simple/Setup.hs#L122

23Skidoo commented:
Merged, thanks!

grayjay commented:
Thanks!

> The right place for BooleanFlag is probably in Distribution.Simple.Setup, near the Flag type:
https://github.com/haskell/cabal/blob/master/Cabal/Distribution/Simple/Setup.hs#L122

Would you like me to move it?

23Skidoo commented:
> Would you like me to move it?

If it's not too much trouble. It's probably better to do that right now than later.

grayjay commented:
I opened  #3379.
